src/upstream_service.cpp

```diff
@@ -0,0 +1,294 @@
+  if (bytes_parsed != sampling_mechanism_string.size()) {
+    std::ostringstream error;
+    error << "Unable to parse a sampling mechanism integer from the following text (consumed "
+          << bytes_parsed << " bytes out of " << sampling_mechanism_string.size()
+          << "): " << sampling_mechanism_string;
+    throw std::invalid_argument(error.str());
+  }
+  return sampling_mechanism_int;
+} catch (const std::logic_error& error) {
+  throw std::invalid_argument(std::string("Unable to parse sampling mechanism (error: ") +
+                              error.what() + ") from the following text: " + std::string(text));
+}
+
+double deserializeSamplingRate(ot::string_view text) try {
+  if (text.empty()) {
+    return std::nan("");
+  }
+
+  const std::string sampling_rate_string = text;
+  std::size_t bytes_parsed = 0;
+  const double sampling_rate = std::stod(sampling_rate_string, &bytes_parsed);
+  if (bytes_parsed != sampling_rate_string.size()) {
+    std::ostringstream error;
+    error << "Unable to parse a sampling rate float from the following text (consumed "
+          << bytes_parsed << " bytes out of " << sampling_rate_string.size()
+          << "): " << sampling_rate_string;
+    throw std::invalid_argument(error.str());
+  }
+
+  return sampling_rate;
+} catch (const std::logic_error& error) {
+  throw std::invalid_argument(std::string("Unable to parse sampling rate (error: ") +
+                              error.what() + ") from the following text: " + std::string(text));
+}
+
+// Return an `UpstreamService` parsed from the specified `text`, or throw a
+// `std::invalid_argument` if an error occurs.
+UpstreamService deserialize(ot::string_view text) {
+  UpstreamService result;
+  const auto end = text.end();
+
+  // .service_name
+  auto iter = text.begin();
+  auto next = std::find(iter, end, '|');
+  deserializeServiceName(result.service_name, range(iter, next));
+
+  // `UpstreamService` fields are "|" separated.  Every time we go to the next
+  // field, we have to make sure that there was a "|" before it (otherwise
+  // there aren't enough fields).
+  const auto advanceTo = [&](const char* field_name_for_diagnostic) {
+    if (next == end) {
+      std::string error;
+      error += "Missing fields in UpstreamService; no ";
+      error += field_name_for_diagnostic;
+      error += " to decode. Offending text: ";
+      error += text;
+      throw std::invalid_argument(error);
+    }
+    iter = next + 1;
+    next = std::find(iter, end, '|');
+  };
+
+  advanceTo("sampling_priority");
+  result.sampling_priority = deserializeSamplingPriority(range(iter, next));
+
+  advanceTo("sampling_mechanism");
+  result.sampling_mechanism = deserializeSamplingMechanism(range(iter, next));
+
+  advanceTo("sampling_rate");
+  result.sampling_rate = deserializeSamplingRate(range(iter, next));
+
+  // Put any remaining fields into `result.unknown_fields`.
```

Comment:
Caleb: As an alternative to enforcing "additional fields are still pipe separated," we can instead just stop parsing here and store any remaining suffix, which we would later append when reserializing.